packages/es-client/search.js

```diff
@@ -91,29 +111,148 @@ const esMetricsConfig = () => {
   };
 };
 
+/**
+ * Generates a configuration for Elasticsearch based on the environment
+ * and provided parameters.
+ *
+ * @param {string} [host] - The host URL for the Elasticsearch instance.
+ * @param {boolean} [metrics=false] - A flag indicating whether metrics are enabled.
+ * @returns {Promise<[Object, import('@aws-sdk/types').Credentials | undefined]>} A
+ * promise that resolves to a tuple containing the configuration object and
+ * AWS credentials (if applicable).
+ */
 const esConfig = async (host, metrics = false) => {
   let config;
+  let credentials;
   if (inTestMode() || 'LOCAL_ES_HOST' in process.env) {
     config = esTestConfig();
   } else if (metrics) {
     config = esMetricsConfig();
   } else {
-    config = await esProdConfig(host);
+    credentials = await getAwsCredentials();
+    config = esProdConfig(host, credentials);
   }
-  return config;
+  return [config, credentials];
 };
 
+/**
+ * `EsClient` is a class for managing an Elasticsearch client.
+ *
+ * @property {string} host - The host URL for the Elasticsearch instance.
+ * @property {boolean} metrics - A flag indicating whether metrics are enabled.
+ * @property {elasticsearch.Client} _client - The Elasticsearch client instance.
+ *
+ * @method constructor - Initializes a new instance of the `EsClient` class.
+ * @method initializeEsClient - Initializes the Elasticsearch client (this._client/client)
+ * if it hasn't been initialized yet.
+ * @method refreshClient - Refreshes the Elasticsearch client if the AWS credentials have changed,
+ * by creating a new Elasticsearch `Client` instance.
+ * @method client - Getter that returns the current Elasticsearch Client
+ */
+class EsClient {
+  /**
+   *  Initializes the Elasticsearch client if it hasn't been initialized yet,
+   * fetching AWS credentials if necessary.
+   *
+   * @returns {Promise<elasticsearch.Client>} A promise that resolves to an instance of
+   * `elasticsearch.Client`.
+   */
+  async initializeEsClient() {
+    /** @type {elasticsearch.Client | undefined} */
+    let client = this._client;
+    if (!client) {
+      const [config, credentials] = await esConfig(this.host, this.metrics);
+      if (credentials) {
+        this._awsKeyId = credentials.accessKeyId;
+      }
+      client = new elasticsearch.Client(config);
+      this._client = client;
+    }
+    return client;
+  }
+
+  /**
+   * Asynchronously refreshes the Elasticsearch client if the AWS credentials have changed,
+   * by creating a new Elasticsearch `Client` instance.
+   *
+   * @returns {Promise<void>} A promise that resolves when the credentials have been refreshed.
+   */
+  async refreshClient() {
+    const { host, metrics } = this;
+    if (this.metrics || inTestMode() || process.env.LOCAL_ES_HOST) {
+      return;
+    }
+    const oldKey = this._awsKeyId;
+    const newCreds = await getAwsCredentials();
+    if (oldKey !== newCreds.accessKeyId) {
```

Comment:
Could we not do this? We're using `refreshClient` in our indexers. In es-client/indexer it's called after we get an error. In the `indexModel` lambda function it's called before trying to index. Could we move that Lambda call to the `catch` block? Then I think we could skip this check to see if the keys don't match and just always get new credentials?

If that doesn't work for whatever reason, I can't say there's a specific problem here. I winced at referencing the access key directly in code but I can't think of a good reason not to. Really this would just be simplifying if possible.

Reply:
Discussed 4/18. There really isn't a _concern_ checking the keys like this, more of a question about complexity. There are alternatives here but given the somewhat narrowly-scoped issue we're trying to solve (long-running requests where credential time-out mid-way) and knowing that this whole class will be going away in the near future (right??????), this implementation is fine. Anything else would require a not-insignificant amount of effort.